### ISSUE TITLE
save_model/load_model pair does not preserve model's metric

### ISSUE THREAD

**Issue by eli-osherovich** [tensorflow/tensorflow#45573]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian testing
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
After saving and then loading a model it produces different metric results.

**Describe the expected behavior**
Everything should be the same 

**Standalone code to reproduce the issue**
Please see a colab notebook [here](https://colab.research.google.com/drive/1-vUEU1gFf7GRrhy95E-h6xw_UB80yDSB?usp=sharing)

Note that the metric of the loaded model is `accuracy: 1.0`, whereas original model produced `accuracy: 0.125'

**Comment by eli-osherovich** [comment]
Please note also that in this case I do get same loss for both models, however, in my original (more complex) case even the losses are different.

**Comment by bhack** [comment]
I think it could be related to https://github.com/tensorflow/tensorflow/pull/41701.
To verify can you pass an explicit metric name e.g. `metrics=["sparse_categorical_accuracy"]` or another explicit one in your colab?

**Comment by eli-osherovich** [comment]
@bhack 
It is explicit already ('metic'), do you want to change it to another one?

**Comment by bhack** [comment]
As mentioned in the reference ticket I meant "explicit" string in the context of:

> When you pass the strings 'accuracy' or 'acc', we convert this to one of tf.keras.metrics.BinaryAccuracy, tf.keras.metrics.CategoricalAccuracy, tf.keras.metrics.SparseCategoricalAccuracy based on the loss function used and the model output shape. We do a similar conversion for the strings 'crossentropy' and 'ce' as well.

So to pass an explicit metric strings not an automatically selected like "accurancy" or "acc".

**Comment by eli-osherovich** [comment]
It seems, that being more specific (`metrics=["sparse_categorical_accuracy"]`) solves the problem. 
This is strange since the loss is always (both for "acc" and "sparse_categorical_accuracy") is `keras.losses.SparseCategoricalCrossentropy`

I updated the  [notebook](https://colab.research.google.com/drive/1-vUEU1gFf7GRrhy95E-h6xw_UB80yDSB?usp=sharing)

**Comment by bhack** [comment]
What do you mean that both loss and accurancy are wrong?

**Comment by bhack** [comment]
Also try your colab with:

`pip install tf-nightly`

And `tf.random.set_seed(1234)` just after your import

**Comment by eli-osherovich** [comment]
@bhack .
Please have a look at the notebook I posted. By "wrong" I meant that the numbers do not correspond to those printed during training (I know these may be different).

**Comment by bhack** [comment]
> @bhack .
> Please have a look at the notebook I posted. By "wrong" I meant that the numbers do not correspond to those printed during training (I know these may be different).
> 

Yes about this you cannot expect train/fit logs and evaluate with the same loss/accurancy values.

The old long thread that explain this is at https://github.com/keras-team/keras/issues/6977 and also in others Stackoverflow questions/Github issues

Other than this can you try your colab with my previous feedback?

**Comment by eli-osherovich** [comment]
In fact, I look at validation loss before and after save/reload -- I see no reason why these should be different.

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported, please find the gist here for[ tf 2.3](https://colab.research.google.com/gist/Saduf2019/7ccec13796c777a682ace461be587c9b/untitled487.ipynb) and[ nightly](https://colab.research.google.com/gist/Saduf2019/9327e063a6bb9934944a8b9c2877bbda/untitled487.ipynb)

**Comment by byronyi** [comment]
Could you try compile your model after loading it back?

**Comment by byronyi** [comment]
See https://github.com/keras-team/keras/issues/14231

**Comment by eli-osherovich** [comment]
1. According to my understanding, `load_model` compiles the model by default [documentaion](https://www.tensorflow.org/api_docs/python/tf/keras/models/load_model)
2. I tried to compile -- it does not help (as expected)
3. The notebook is available to everyone, hence, I assume that everyone (including developers) can play with compilation, evaluation sets, etc. Am I getting this wrong?
4.

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/9f1214a0547dbed6dba81f4eb364d528/untitled301.ipynb) ..Thanks!

**Comment by bhack** [comment]
@sushreebarsa I cannot reproduce with `tf.random.set_seed(1234)` in your colab

**Comment by jvishnuvardhan** [comment]
@eli-osherovich 
Looks like this was resolved in recent TF versions. [Here](https://colab.research.google.com/gist/jvishnuvardhan/75fe1b92033caa59c6114900536e766c/save_load_metrics_bug.ipynb) is a gist for reference.

Can you please verify once and close the issue if this was resolved for you.

If this was not resolved, can you please open the issue in keras-team/keras repo repo as keras development moved to that repo to focus mainly on Keras. Thanks!

**Comment by eli-osherovich** [comment]
@jvishnuvardhan 
Is there a unit test that checks it? If not, this is just a coincidence.

**Comment by bhack** [comment]
> @jvishnuvardhan 
> Is there a unit test that checks it? If not, this is just a coincidence.
> 

It you can extend a test with a PR to cover this It could be useful

**Comment by jvishnuvardhan** [comment]
I will raise a PR to add a unit test. I am closing this issue as this was resolved. Thanks!
